web-common/src/features/canvas/components/BaseCanvasComponent.ts

```diff
@@ -0,0 +1,88 @@
+import type { ChartType } from "@rilldata/web-common/features/canvas/components/charts/types";
+import type {
+  CanvasComponent,
+  ComponentSize,
+} from "@rilldata/web-common/features/canvas/components/types";
+import { getParsedDocument } from "@rilldata/web-common/features/canvas/inspector/selectors";
+import type { InputParams } from "@rilldata/web-common/features/canvas/inspector/types";
+import type { FileArtifact } from "@rilldata/web-common/features/entity-management/file-artifact";
+import { get, writable, type Writable } from "svelte/store";
+
+// A base class that implements all the store logic
+export abstract class BaseCanvasComponent<T> implements CanvasComponent<T> {
```

Comment:
Feels a little weird having an interface and an abstract class. We should have a single layer.

Another options is to move to `has-a` relation with child components being a member. These classes will only have minSize, defaultSize, isValid, inputParams and newComponentSpec.